orttraining/orttraining/core/optimizer/memory_optimizer.cc

```diff
@@ -0,0 +1,775 @@
+
+      // Skip if the node output is already visited.
+      if (std::find(visited_output_arg_set.begin(), visited_output_arg_set.end(), p) !=
+          visited_output_arg_set.end()) {
+        continue;
+      }
+
+      visited_output_arg_set.insert({p});
+
+      // If the node already visited by from it's other output index, skip it.
+      if (visited_node_set.find(curr_node) != visited_node_set.end()) {
+        continue;
+      }
+
+      visited_node_set.insert(curr_node);
+
+      // Buttom-up search rules.
+      // If current op is in allowed list, check its input args, and append the producers' NodeOutputPorts to next_q.
+      // If current op is NOT in allowed list:
+      // 1). the output does not exist in backward, we cannot find a good solution for so, search terminates.
+      // 2). the output is used in backward, we don't need trace back further, continue searching.
+      auto op_recompute_config_it = recomputable_op_type_to_input_arg_index_map_.find(curr_node->OpType());
+      auto cur_output_arg_name = curr_node->OutputDefs()[p.second]->Name();
+      if (is_entry_node_output_arg) {
+        // We handle the entry node outputs differently because, we don't want this case falls into and succeed one of
+        // the checks in the other branch
+        // 1. "op is not in recompute op list, but its output is used in backward"
+        // 2. "op is in recompute op list, but its output is used in backward"
+        // (either of the above checks is true for entry node outputs)
+        if (op_recompute_config_it == recomputable_op_type_to_input_arg_index_map_.end()) {
+          early_stop = true;
+          LOGS(logger, VERBOSE) << "Entry Node " << curr_node->Name() << "(" << curr_node->OpType() << ") is **NOT** "
+                                << "in recompute op list, search terminates.";
+          break;
+        }
+      } else {
+        if (op_recompute_config_it == recomputable_op_type_to_input_arg_index_map_.end()) {
+          if (fw_op_output_arg_used_map.at(cur_output_arg_name).second) {
+            LOGS(logger, VERBOSE) << "Node " << curr_node->Name() << "(" << curr_node->OpType() << ") is **NOT** in "
+                                  << "recompute op list, but its output [" << cur_output_arg_name
+                                  << "] is used in backward, we don't need trace buttom-up further";
+            continue;
+          } else {
+            early_stop = true;
+            LOGS(logger, VERBOSE) << "Node " << curr_node->Name() << "(" << curr_node->OpType() << ") is **NOT** in "
+                                  << "recompute op list, and its output [" << cur_output_arg_name
+                                  << "] does not exist in backward, search terminates.";
+            break;
+          }
+        }
+
+        if (fw_op_output_arg_used_map.at(cur_output_arg_name).second) {
+          LOGS(logger, VERBOSE) << "Node " << curr_node->Name() << "(" << curr_node->OpType() << ") "
```

Comment:
Oh do you mean, we should mention it like "we don't need trace bottom-up further in current path"? Or let me know if I misunderstood it.

Reply:
yeah, that is what I meant. From the logs, it will look like that node name is not needed to be traversed. An alternative is to add a message for entry nodes, like `"Starting search from node " << entry_node->Name();`. This way users will know the current path.